### PR TITLE
added new enteties shopping cart and item cart, also create new contr…

### PR DESCRIPTION
**Pull request by Dima23-ops5**
…oller

### CODE REVIEW

File: src/main/java/mate/academy/springbootintro/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
+    @GetMapping
+    @PreAuthorize(value = "hasRole('USER')")
+    @Operation(
+            summary = "Get shopping cart",
+            description = "Getting shopping cart for current user"
+    )
+    public ShoppingCardDto getShoppingCard() {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCardDto getShoppingCard(Authentication authentication) {
```

File: src/main/java/mate/academy/springbootintro/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    @DeleteMapping(value = "/items/{cartItemId}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @PreAuthorize(value = "hasRole('USER')")
+    @Operation(summary = "Delete item cart", description = "Deleting item carts in shopping cart")
+    public void deleteItemCart(Long cartItemId) {
+        shoppingCartService.deleteCartItem(cartItemId);
+    }
+
+    private Long getCurrentUserId() {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/controller/ShoppingCartController.java]:
```suggestion
```

File: src/main/java/mate/academy/springbootintro/controller/ShoppingCartController.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+    @DeleteMapping(value = "/items/{cartItemId}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @PreAuthorize(value = "hasRole('USER')")
+    @Operation(summary = "Delete item cart", description = "Deleting item carts in shopping cart")
+    public void deleteItemCart(Long cartItemId) {
+        shoppingCartService.deleteCartItem(cartItemId);
+    }
+
+    private Long getCurrentUserId() {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/controller/ShoppingCartController.java]:
```suggestion
    private Long getCurrentUserId(Authentication authentication) {
```

File: src/main/java/mate/academy/springbootintro/controller/ShoppingCartController.java (hunk 1 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/springbootintro/controller/ShoppingCartController.java]: same for other endpoints

File: src/main/java/mate/academy/springbootintro/dto/cartitem/CreateCartItemRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package mate.academy.springbootintro.dto.cartitem;
+
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.Positive;
+
+public record CreateCartItemRequestDto(
+        @NotBlank @Positive Long bookId,
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/dto/cartitem/CreateCartItemRequestDto.java]: NotBlank is used only for String
```suggestion
        @NotNull @Positive Long bookId,
```

File: src/main/java/mate/academy/springbootintro/dto/cartitem/CreateCartItemRequestDto.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+        @NotBlank int quantity
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/dto/cartitem/CreateCartItemRequestDto.java]:
```suggestion
        @NotNull @Positive int quantity
```

File: src/main/java/mate/academy/springbootintro/dto/cartitem/UpdateCartItemRequestDto.java [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package mate.academy.springbootintro.dto.cartitem;
+
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.Positive;
+
+public record UpdateCartItemRequestDto(
+        @NotBlank @Positive int quantity
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/dto/cartitem/UpdateCartItemRequestDto.java]:
```suggestion
        @NotNull @Positive int quantity
```

File: src/main/java/mate/academy/springbootintro/model/CartItem.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+@Table(name = "cart_items")
+@Setter
+@Getter
+@SQLDelete(sql = "UPDATE cart_items SET is_deleted = TRUE WHERE id = ?")
+@SQLRestriction(value = "is_deleted = FALSE")
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/mate/academy/springbootintro/model/CartItem.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    @JoinColumn(name = "shoppingCart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+    @ManyToOne(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/mate/academy/springbootintro/model/ShoppingCart.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+@SQLDelete(sql = "UPDATE shopping_cards SET is_deleted = TRUE WHERE id = ?")
+@SQLRestriction(value = "is_deleted = FALSE")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "user_id", nullable = false)
+    private User user;
+    @OneToMany(mappedBy = "shoppingCart")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/model/ShoppingCart.java]:
```suggestion
    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/java/mate/academy/springbootintro/service/shoppingcart/impl/ShoppingCartServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+        );
+        Book book = bookRepository.findById(
+                cartItemRequestDto.bookId()
+        ).orElseThrow(
+                () -> new EntityNotFoundException(
+                        "Cannot found book with id: "
+                                + cartItemRequestDto.bookId()
+                )
+        );
+        CartItem cartItem = cartItemRepository.findByShoppingCartAndBook(
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/service/shoppingcart/impl/ShoppingCartServiceImpl.java]:
```suggestion
        shoppingCart.getCartItems().stream()
                .filter(item -> item.getBook().getId().equals(cartItemRequestDto.bookId()))
                .findFirst()
                .ifPresentOrElse(item -> item.setQuantity(item.getQuantity() + cartItemRequestDto.quantity()),
                        () -> addCartItemToCart(cartItemRequestDto, book, cart));
        shoppingCartRepository.save(shoppingCart);
        return shoppingCardMapper.toDto(shoppingCart);
```

Where
```
private void addCartItemToCart(CartItemRequestDto itemDto, Book book, ShoppingCart cart) {
        CartItem cartItem = itemMapper.toCartItem(itemDto);
        cartItem.setBook(book);
        cart.addItemToCart(cartItem);
    }
```

Where in ShoppingCart
```
public void addItemToCart(CartItem cartItem) {
        cartItem.setShoppingCart(this);
        cartItems.add(cartItem);
    }
```

File: src/main/java/mate/academy/springbootintro/dto/shoppingcartdto/ShoppingCardDto.java [hunk 11]
```diff
@@ -0,0 +1,11 @@
+package mate.academy.springbootintro.dto.shoppingcartdto;
+
+import java.util.Set;
+import mate.academy.springbootintro.model.CartItem;
+
+public record ShoppingCardDto(
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/dto/shoppingcartdto/ShoppingCardDto.java]:
```suggestion
public record ShoppingCartDto(
```

File: src/main/java/mate/academy/springbootintro/mapper/ShoppingCardMapper.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
+package mate.academy.springbootintro.mapper;
+
+import mate.academy.springbootintro.config.MapperConfig;
+import mate.academy.springbootintro.dto.shoppingcartdto.ShoppingCardDto;
+import mate.academy.springbootintro.model.ShoppingCart;
+import org.mapstruct.Mapper;
+
+@Mapper(config = MapperConfig.class)
+public interface ShoppingCardMapper {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/mapper/ShoppingCardMapper.java]:
```suggestion
public interface ShoppingCartMapper {
```

File: src/main/java/mate/academy/springbootintro/service/shoppingcart/impl/ShoppingCartServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+        );
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCardDto updateQuantityInCartItem(
+            Long cartItemId,
+            UpdateCartItemRequestDto updateCartItemRequestDto,
+            Long currentUserId) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/service/shoppingcart/impl/ShoppingCartServiceImpl.java]: Suggestion

```suggestion
        ShoppingCart cart = repository.findByUserId(userId);
        CartItem cartItem = itemRepository.findByIdAndShoppingCartId(cartItemId, cart.getId())
                .map(item -> {
                    item.setQuantity(itemDto.quantity());
                    return item;
                }).orElseThrow(() -> new EntityNotFoundException(
                        String.format("No cart item with id: %d for user: %d", cartItemId, userId)
                ));
        itemRepository.save(cartItem);
        return cartMapper.toDto(cart);
```

File: src/main/java/mate/academy/springbootintro/service/shoppingcart/impl/ShoppingCartServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+        cartItem.setQuantity(updateCartItemRequestDto.quantity());
+        cartItemRepository.save(cartItem);
+        shoppingCart.getCartItems().add(cartItem);
+
+        return shoppingCardMapper.toDto(shoppingCartRepository.save(shoppingCart));
+    }
+
+    @Override
+    public void deleteCartItem(Long cartItemId) {
+        CartItem cartItem = cartItemRepository.findById(
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/service/shoppingcart/impl/ShoppingCartServiceImpl.java]:
```suggestion
        ShoppingCart shoppingCart = repository.findByUserId(userId);
        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(
```

File: src/main/java/mate/academy/springbootintro/repository/cartitem/CartItemRepository.java [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+import java.util.Optional;
+import mate.academy.springbootintro.model.Book;
+import mate.academy.springbootintro.model.CartItem;
+import mate.academy.springbootintro.model.ShoppingCart;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface CartItemRepository extends JpaRepository<CartItem, Long> {
+    Optional<CartItem> findByShoppingCartAndBook(ShoppingCart shoppingCart, Book book);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/repository/cartitem/CartItemRepository.java]:
```suggestion
    Optional<CartItem> findByIdAndShoppingCartId(Long itemId, Long cartId);
```

File: src/main/java/mate/academy/springbootintro/repository/shoppingcart/ShoppingCartRepository.java [hunk 16]
```diff
@@ -0,0 +1,11 @@
+package mate.academy.springbootintro.repository.shoppingcart;
+
+import java.util.Optional;
+import mate.academy.springbootintro.model.ShoppingCart;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+    Optional<ShoppingCart> findByUserId(Long userId);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/repository/shoppingcart/ShoppingCartRepository.java]:
```suggestion
    @EntityGraph(attributePaths = {"cartItems", "cartItems.book"})
    ShoppingCart findByUserId(Long userId);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/java/mate/academy/springbootintro/controller/BookController.java [hunk 17]
```diff
@@ -39,6 +40,7 @@ public class BookController {
 
     @GetMapping
     @Operation(summary = "Get books", description = "Get all books")
+    @PreAuthorize(value = "hasRole('USER')")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/controller/BookController.java]:
```suggestion
    @PreAuthorize("hasRole('USER')")
```

check other places

File: src/main/java/mate/academy/springbootintro/controller/ShoppingCartController.java [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+    @DeleteMapping(value = "/items/{cartItemId}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @PreAuthorize(value = "hasRole('USER')")
+    @Operation(summary = "Delete item cart", description = "Deleting item carts in shopping cart")
+    public void deleteItemCart(Long cartItemId, Authentication authentication) {
+        shoppingCartService.deleteCartItem(cartItemId, getCurrentUserId(authentication));
+    }
+
+    private Long getCurrentUserId(Authentication authentication) {
+        authentication = SecurityContextHolder.getContext().getAuthentication();
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/controller/ShoppingCartController.java]:
```suggestion
```

File: src/main/java/mate/academy/springbootintro/controller/ShoppingCartController.java [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/controller/ShoppingCartController.java:28]:
```suggestion
@RestController
@Validated
```

File: src/main/java/mate/academy/springbootintro/dto/shoppingcartdto/ShoppingCardDto.java (hunk 11 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/springbootintro/dto/shoppingcartdto/ShoppingCardDto.java]: Not fixed

File: src/main/java/mate/academy/springbootintro/mapper/ShoppingCardMapper.java (hunk 12 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/springbootintro/mapper/ShoppingCardMapper.java]: Not fixed

File: src/main/java/mate/academy/springbootintro/model/CartItem.java (hunk 7 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/springbootintro/model/CartItem.java]: not fixed

File: src/main/java/mate/academy/springbootintro/model/CartItem.java (hunk 8 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/springbootintro/model/CartItem.java]: not fixed

File: src/main/java/mate/academy/springbootintro/model/ShoppingCart.java (hunk 9 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/springbootintro/model/ShoppingCart.java]: not fixed

File: src/main/java/mate/academy/springbootintro/repository/cartitem/CartItemRepository.java (hunk 15 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/springbootintro/repository/cartitem/CartItemRepository.java]: not fixed

File: src/main/java/mate/academy/springbootintro/repository/shoppingcart/ShoppingCartRepository.java (hunk 16 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/springbootintro/repository/shoppingcart/ShoppingCartRepository.java]: not fixed

File: src/main/java/mate/academy/springbootintro/service/shoppingcart/impl/ShoppingCartServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+    private final BookRepository bookRepository;
+    private final CartItemRepository cartItemRepository;
+    private final ShoppingCardMapper shoppingCardMapper;
+
+    @Override
+    @Transactional
+    public ShoppingCardDto findShoppingCardByUserId(Long id) {
+        return shoppingCardMapper.toDto(
+                shoppingCartRepository.findByUserId(id)
+                        .orElseThrow(
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/service/shoppingcart/impl/ShoppingCartServiceImpl.java]: reformat your style, it's hard to read your code (all places)

File: src/main/java/mate/academy/springbootintro/service/shoppingcart/impl/ShoppingCartServiceImpl.java (hunk 10 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/springbootintro/service/shoppingcart/impl/ShoppingCartServiceImpl.java]: not fixed

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Fix previous comments. If you have some questions, please ask in chat

File: src/main/java/mate/academy/springbootintro/controller/ShoppingCartController.java [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
+    @GetMapping
+    @PreAuthorize(value = "hasRole('USER')")
+    @Operation(
+            summary = "Get shopping cart",
+            description = "Getting shopping cart for current user"
+    )
+    public ShoppingCartDto getShoppingCard(Authentication authentication) {
```
Comment (okuzan) [inline src/main/java/mate/academy/springbootintro/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartDto getShoppingCart(Authentication authentication) {
```

File: src/main/java/mate/academy/springbootintro/dto/cartitem/CreateCartItemRequestDto.java [hunk 22]
```diff
@@ -0,0 +1,10 @@
+package mate.academy.springbootintro.dto.cartitem;
+
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Positive;
+
+public record CreateCartItemRequestDto(
+        @NotNull @Positive Long bookId,
+        @NotNull int quantity
```
Comment (okuzan) [inline src/main/java/mate/academy/springbootintro/dto/cartitem/CreateCartItemRequestDto.java]:
```suggestion
        @Positive int quantity
```

File: src/main/java/mate/academy/springbootintro/dto/cartitem/UpdateCartItemRequestDto.java [hunk 23]
```diff
@@ -0,0 +1,9 @@
+package mate.academy.springbootintro.dto.cartitem;
+
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Positive;
+
+public record UpdateCartItemRequestDto(
+        @NotNull @Positive int quantity
```
Comment (okuzan) [inline src/main/java/mate/academy/springbootintro/dto/cartitem/UpdateCartItemRequestDto.java]:
```suggestion
          @Positive int quantity
```
primitive can't be null

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/springbootintro/controller/ShoppingCartController.java (hunk 19 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/springbootintro/controller/ShoppingCartController.java:28]: `@Positive Long id` annotation won't work without `@Validated`

File: src/main/java/mate/academy/springbootintro/controller/ShoppingCartController.java (hunk 18 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/springbootintro/controller/ShoppingCartController.java]: Why you need this ?
`authentication = SecurityContextHolder.getContext().getAuthentication();`

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by Elena-Bruyako — APPROVED** [review]
Well done

**Pull request closed without merging**